// winston/lib/winston.js
import __IMPORT_11__ from '/web/node_modules/winston/lib/winston.js'
import __IMPORT_10__ from '/web/node_modules/winston-transport/index.js'
import __IMPORT_9__ from '/web/node_modules/winston/lib/winston/container.js'
import __IMPORT_8__ from '/web/node_modules/winston/lib/winston/rejection-handler.js'
import __IMPORT_7__ from '/web/node_modules/winston/lib/winston/exception-handler.js'
import __IMPORT_6__ from '/web/node_modules/winston/lib/winston/logger.js'
import __IMPORT_5__ from '/web/node_modules/winston/lib/winston/create-logger.js'
import __IMPORT_4__ from '/web/node_modules/winston/lib/winston/config/index.js'
import __IMPORT_3__ from '/web/node_modules/winston/lib/winston/transports/index.js'
import __IMPORT_2__ from '/web/node_modules/winston/package.js'
import __IMPORT_1__ from '/web/node_modules/winston/lib/winston/common.js'
import __IMPORT_0__ from '/web/node_modules/logform/index.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
/**
 * winston.js: Top-level include defining Winston.
 *
 * (C) 2010 Charlie Robbins
 * MIT LICENCE
 */

'use strict';

const logform = __IMPORT_0__();
const { warn } = __IMPORT_1__();

/**
 * Expose version. Use `require` method for `webpack` support.
 * @type {string}
 */
exports.version = __IMPORT_2__().version;
/**
 * Include transports defined by default by winston
 * @type {Array}
 */
exports.transports = __IMPORT_3__();
/**
 * Expose utility methods
 * @type {Object}
 */
exports.config = __IMPORT_4__();
/**
 * Hoist format-related functionality from logform.
 * @type {Object}
 */
exports.addColors = logform.levels;
/**
 * Hoist format-related functionality from logform.
 * @type {Object}
 */
exports.format = logform.format;
/**
 * Expose core Logging-related prototypes.
 * @type {function}
 */
exports.createLogger = __IMPORT_5__();
/**
 * Expose core Logging-related prototypes.
 * @type {function}
 */
exports.Logger = __IMPORT_6__();
/**
 * Expose core Logging-related prototypes.
 * @type {Object}
 */
exports.ExceptionHandler = __IMPORT_7__();
/**
 * Expose core Logging-related prototypes.
 * @type {Object}
 */
exports.RejectionHandler = __IMPORT_8__();
/**
 * Expose core Logging-related prototypes.
 * @type {Container}
 */
exports.Container = __IMPORT_9__();
/**
 * Expose core Logging-related prototypes.
 * @type {Object}
 */
exports.Transport = __IMPORT_10__();
/**
 * We create and expose a default `Container` to `winston.loggers` so that the
 * programmer may manage multiple `winston.Logger` instances without any
 * additional overhead.
 * @example
 *   // some-file1.js
 *   const logger = __IMPORT_11__().loggers.get('something');
 *
 *   // some-file2.js
 *   const logger = __IMPORT_11__().loggers.get('something');
 */
exports.loggers = new exports.Container();

/**
 * We create and expose a 'defaultLogger' so that the programmer may do the
 * following without the need to create an instance of winston.Logger directly:
 * @example
 *   const winston = __IMPORT_11__();
 *   winston.log('info', 'some message');
 *   winston.error('some error');
 */
const defaultLogger = exports.createLogger();

// Pass through the target methods onto `winston.
Object.keys(exports.config.npm.levels)
  .concat([
    'log',
    'query',
    'stream',
    'add',
    'remove',
    'clear',
    'profile',
    'startTimer',
    'handleExceptions',
    'unhandleExceptions',
    'handleRejections',
    'unhandleRejections',
    'configure',
    'child'
  ])
  .forEach(
    method => (exports[method] = (...args) => defaultLogger[method](...args))
  );

/**
 * Define getter / setter for the default logger level which need to be exposed
 * by winston.
 * @type {string}
 */
Object.defineProperty(exports, 'level', {
  get() {
    return defaultLogger.level;
  },
  set(val) {
    defaultLogger.level = val;
  }
});

/**
 * Define getter for `exceptions` which replaces `handleExceptions` and
 * `unhandleExceptions`.
 * @type {Object}
 */
Object.defineProperty(exports, 'exceptions', {
  get() {
    return defaultLogger.exceptions;
  }
});

/**
 * Define getter for `rejections` which replaces `handleRejections` and
 * `unhandleRejections`.
 * @type {Object}
 */
Object.defineProperty(exports, 'rejections', {
  get() {
    return defaultLogger.rejections;
  }
});

/**
 * Define getters / setters for appropriate properties of the default logger
 * which need to be exposed by winston.
 * @type {Logger}
 */
['exitOnError'].forEach(prop => {
  Object.defineProperty(exports, prop, {
    get() {
      return defaultLogger[prop];
    },
    set(val) {
      defaultLogger[prop] = val;
    }
  });
});

/**
 * The default transports and exceptionHandlers for the default winston logger.
 * @type {Object}
 */
Object.defineProperty(exports, 'default', {
  get() {
    return {
      exceptionHandlers: defaultLogger.exceptionHandlers,
      rejectionHandlers: defaultLogger.rejectionHandlers,
      transports: defaultLogger.transports
    };
  }
});

// Have friendlier breakage notices for properties that were exposed by default
// on winston < 3.0.
warn.deprecated(exports, 'setLevels');
warn.forFunctions(exports, 'useFormat', ['cli']);
warn.forProperties(exports, 'useFormat', ['padLevels', 'stripColors']);
warn.forFunctions(exports, 'deprecated', [
  'addRewriter',
  'addFilter',
  'clone',
  'extend'
]);
warn.forProperties(exports, 'deprecated', ['emitErrs', 'levelLength']);


return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
