// readable-stream/readable.js
import __IMPORT_7__ from '/web/node_modules/readable-stream/lib/internal/streams/pipeline.js'
import __IMPORT_6__ from '/web/node_modules/readable-stream/lib/internal/streams/end-of-stream.js'
import __IMPORT_5__ from '/web/node_modules/readable-stream/lib/_stream_passthrough.js'
import __IMPORT_4__ from '/web/node_modules/readable-stream/lib/_stream_transform.js'
import __IMPORT_3__ from '/web/node_modules/readable-stream/lib/_stream_duplex.js'
import __IMPORT_2__ from '/web/node_modules/readable-stream/lib/_stream_writable.js'
import __IMPORT_1__ from '/web/node_modules/readable-stream/lib/_stream_readable.js'
import __IMPORT_0__ from '/web/node/stream.mjs'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
var Stream = __IMPORT_0__;
if (process.env.READABLE_STREAM === 'disable' && Stream) {
  module.exports = Stream.Readable;
  Object.assign(module.exports, Stream);
  module.exports.Stream = Stream;
} else {
  exports = module.exports = __IMPORT_1__();
  exports.Stream = Stream || exports;
  exports.Readable = exports;
  exports.Writable = __IMPORT_2__();
  exports.Duplex = __IMPORT_3__();
  exports.Transform = __IMPORT_4__();
  exports.PassThrough = __IMPORT_5__();
  exports.finished = __IMPORT_6__();
  exports.pipeline = __IMPORT_7__();
}

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
