// @sap/cds/lib/index.js
import __IMPORT_44__ from 'utils/lazify.js'
import __IMPORT_43__ from 'ql/cds-ql.js'
import __IMPORT_42__ from 'ql/cds.ql-infer.js'
import __IMPORT_41__ from 'log/cds-log.js'
import __IMPORT_40__ from 'test/cds-test.js'
import __IMPORT_39__ from '../bin/serve.js'
import __IMPORT_38__ from 'log/cds-error.js'
import __IMPORT_37__ from 'utils/cds-utils.js'
import __IMPORT_36__ from '../srv/remote-service.js'
import __IMPORT_35__ from '../srv/db-service.js'
import __IMPORT_34__ from '../srv/msg-service.js'
import __IMPORT_33__ from '../srv/app-service.js'
import __IMPORT_32__ from 'req/validate.js'
import __IMPORT_31__ from 'req/user.js'
import __IMPORT_30__ from 'req/event.js'
import __IMPORT_29__ from 'req/request.js'
import __IMPORT_28__ from 'req/context.js'
import __IMPORT_27__ from 'srv/cds.Service.js'
import __IMPORT_26__ from 'srv/middlewares/auth/index.js'
import __IMPORT_25__ from '../libx/odata/index.js'
import __IMPORT_24__ from 'srv/middlewares/index.js'
import __IMPORT_23__ from '../libx/queue/index.js'
import __IMPORT_22__ from 'srv/cds-connect.js'
import __IMPORT_21__ from 'srv/cds-serve.js'
import __IMPORT_20__ from '../server.js'
import __IMPORT_19__ from 'core/types.js'
import __IMPORT_18__ from 'core/linked-csn.js'
import __IMPORT_17__ from 'i18n/index.js'
import __IMPORT_16__ from 'i18n/localize.js'
import __IMPORT_15__ from 'dbs/cds-deploy.js'
import __IMPORT_14__ from 'compile/extend.js'
import __IMPORT_13__ from 'compile/minify.js'
import __IMPORT_12__ from 'compile/parse.js'
import __IMPORT_11__ from 'compile/load.js'
import __IMPORT_10__ from 'compile/resolve.js'
import __IMPORT_9__ from 'compile/cds-compile.js'
import __IMPORT_8__ from 'compile/cdsc.js'
import __IMPORT_7__ from 'env/schemas/index.js'
import __IMPORT_6__ from 'env/cds-env.js'
import __IMPORT_5__ from '../package.js'
import __IMPORT_4__ from 'plugins.js'
import __IMPORT_3__ from 'req/spawn.js'
import __IMPORT_2__ from '../../../../node/events.mjs'
import __IMPORT_1__ from '../../../../node/async_hooks.mjs'
import __IMPORT_0__ from 'utils/version.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
if (process.env.CDS_STRICT_NODE_VERSION !== 'false') __IMPORT_0__().check()
! (global.__cds_loaded_from ??= new Set).add(__filename.toLowerCase()) // track from where we loaded cds, lowercase to avoid path duplicates on Windows

const { AsyncLocalStorage } = __IMPORT_1__
const context = new AsyncLocalStorage

const { EventEmitter } = __IMPORT_2__
const cds = exports = module.exports = global.cds = new class cds extends EventEmitter {

  /** Contexts and Transactions @type {import('./req/context.js')} */
  get context()  { return context.getStore() }
  set context(x) { this._with(x) }
  _with (x,fn,...args) {
    const ctx = typeof x !== 'object' || x instanceof exports.EventContext ? x : x.context || exports.EventContext.for(x)
    return fn ? context.run (ctx,fn,...args) : context.enterWith (ctx)
  }
  get spawn() { return super.spawn = __IMPORT_3__() }

  /** @import {LinkedCSN} from './core/linked-csn.js' */
  /** @import {Service} from './srv/cds.Service.js' */
  /** @type LinkedCSN */ model = undefined
  /** @type Service */ db = undefined
  /** CLI args */ cli = { command:'', options:{}, argv:[] }
  /** Working dir */ root = process.cwd()

  emit (eve, ...args) {
    if (eve === 'served') return this.listeners(eve) .reduce (
      (p,each) => p.then(()=> each.call(this,...args)),
      Promise.resolve()
    )
    else return super.emit (eve, ...args)
  }

  // Configuration & Information
  get requires() { return super.requires = this.env.requires._resolved() }
  get plugins()  { return super.plugins = __IMPORT_4__().activate() }
  get version()  { return super.version = __IMPORT_5__().version }
  get env()      { return super.env = __IMPORT_6__().for(this) }
  get home()     { return super.home = __dirname.slice(0,-4) }
  get schema()   { return super.schema = __IMPORT_7__() } // REVISIT: Better move that to cds-dk?

  // Loading and Compiling Models
  get compiler() { return super.compiler = __IMPORT_8__() }
  get compile()  { return super.compile = __IMPORT_9__() }
  get resolve()  { return super.resolve = __IMPORT_10__() }
  get load()     { return super.load = __IMPORT_11__() }
  get get()      { return super.get = this.load.parsed }
  get parse()    { return super.parse = __IMPORT_12__() }
  get minify()   { return super.minify = __IMPORT_13__() }
  get extend()   { return super.extend = __IMPORT_14__() }
  get deploy()   { return super.deploy = __IMPORT_15__() }
  get localize() { return super.localize = __IMPORT_16__() }
  get i18n()     { return super.i18n = __IMPORT_17__() }

  // Model Reflection, Builtin types and classes
  get entities()    { return (this.model||this.db?.model)?.entities }
  get reflect()     { return super.reflect = this.linked }
  get linked()      { return super.linked = __IMPORT_18__() }
  get builtin()     { return super.builtin = __IMPORT_19__() }
  get Association() { return super.Association = this.builtin.classes.Association }
  get Composition() { return super.Composition = this.builtin.classes.Composition }
  get entity()      { return super.entity = this.builtin.classes.entity }
  get event()       { return super.event = this.builtin.classes.event }
  get type()        { return super.type = this.builtin.classes.type }
  get array()       { return super.array = this.builtin.classes.array }
  get struct()      { return super.struct = this.builtin.classes.struct }
  get service()     { return super.service = Object.assign (this.builtin.classes.service, {
    /** @param {( this:Service, srv:Service )} fn */ impl: fn => fn,
    /** @type Service[] */ providers: []
  })}

  // Providing and Consuming Services
  /** @type { Record<string,Service> } */ services = new class {
    *[Symbol.iterator](){ for (let e in this) yield this[e] }
  }
  get server() { return super.server = __IMPORT_20__() }
  get serve() { return super.serve = __IMPORT_21__() }
  get connect() { return super.connect = __IMPORT_22__() }
  get outboxed() { return this.queued }
  get unboxed() { return this.unqueued }
  get queued() { return super.queued = __IMPORT_23__().queued }
  get unqueued() { return super.unqueued = __IMPORT_23__().unqueued }
  get flush() { return super.flush = __IMPORT_23__().cdsFlush }
  get middlewares() { return super.middlewares = __IMPORT_24__() }
  get odata() { return super.odata = __IMPORT_25__() }
  get auth() { return super.auth = __IMPORT_26__() }
  shutdown() { this.app?.server && process.exit() } // is overridden in bin/serve.js

  // Core Services API
  get Service() { return super.Service = __IMPORT_27__() }
  get EventContext() { return super.EventContext = __IMPORT_28__() }
  get Request() { return super.Request = __IMPORT_29__() }
  get Event() { return super.Event = __IMPORT_30__() }
  get User() { return super.User = __IMPORT_31__() }
  get validate() { return super.validate = __IMPORT_32__() }

  // Services, Protocols and Periphery
  get ApplicationService() { return super.ApplicationService = __IMPORT_33__() }
  get MessagingService() { return super.MessagingService = __IMPORT_34__() }
  get DatabaseService() { return super.DatabaseService = __IMPORT_35__() }
  get RemoteService() { return super.RemoteService = __IMPORT_36__() }

  // Helpers
  get utils() { return super.utils = __IMPORT_37__() }
  get error() { return super.error = __IMPORT_38__() }
  get exec() { return super.exec = __IMPORT_39__().exec }
  get test() { return super.test = __IMPORT_40__() }
  get log() { return super.log = __IMPORT_41__() }
  get debug() { return super.debug = this.log.debug }
  clone(x) { return structuredClone(x) }

  // Querying and Databases
  get infer(){ return super.infer = __IMPORT_42__() }
  get txs()  { return super.txs = new this.Service('cds.tx') }
  get ql()   { return super.ql = __IMPORT_43__() }
  // get tx() { return super.tx = this.Service.prototype.tx.bind (new this.Service) } // Note: this would break too much of existing code/tests
  tx         (..._) { return (this.db || this.txs).tx(..._) }
  run        (..._) { return (this.db || typeof _[0] === 'function' && this.txs || this.error._no_primary_db).run(..._) }
  foreach    (..._) { return (this.db || this.error._no_primary_db).foreach(..._) }
  read       (..._) { return this.ql.SELECT(..._) }
  create     (..._) { return this.ql.INSERT.into(..._) }
  insert     (..._) { return this.ql.INSERT(..._) }
  upsert     (..._) { return this.ql.UPSERT(..._) }
  update     (..._) { return this.ql.UPDATE.entity(..._) }
  delete     (..._) { return this.ql.DELETE.from(..._) }
  disconnect (..._) { return this.db?.disconnect(..._) }

  // Deprecated stuff to be removed in upcomming releases...
  /** @deprecated */ get lazified() { return this.lazify }
  /** @deprecated */ get lazify() { return super.lazify = __IMPORT_44__() }
  /** @deprecated */ get in() { return super.in = cwd => !cwd ? this : {__proto__:this, cwd, env: this.env.for('cds',cwd) } }
  /** @deprecated */ exit(code){ return this.app?.server ? this.shutdown() : process.exit(code) }
  /** @deprecated */ transaction (..._) { return (this.db||this.error._no_primary_db).transaction(..._) }
}

// Add global convenience shortcuts for cds.ql and cds.parse commands
cds.extend (global) .with ( class globals {
  get SELECT() { return Q('SELECT') }
  get INSERT() { return Q('INSERT') }
  get UPSERT() { return Q('UPSERT') }
  get UPDATE() { return Q('UPDATE') }
  get DELETE() { return Q('DELETE') }
  /** @deprecated */ get CREATE() { return D('CREATE', '{CREATE} = cds.ql', cds.ql.CREATE) }
  /** @deprecated */ get DROP() { return D('DROP', '{DROP} = cds.ql', cds.ql.DROP) }
  /** @deprecated */ get CDL() { return D('CDL', 'cds.parse.cdl', cds.parse.cdl) }
  /** @deprecated */ get CQL() { return D('CQL', 'cds.parse.cql', cds.parse.cql) }
  /** @deprecated */ get CXL() { return D('CXL', 'cds.parse.expr', cds.parse.expr) }
} .prototype )
function Q (p) { return G (p, cds.ql[p]) }
function D (old,use,fn) { return G (old, cds.utils.deprecated (fn,{kind:'Global constant',old,use})) }
function G (p,v) { Object.defineProperty (global, p, { value:v, enumerable:1,configurable:1}); return v }

// Allow for `import cds from '@sap/cds'` without `esModuleInterop` in tsconfig.json
Object.defineProperties (exports, { default: {value:cds}, __esModule: {value:true} })

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
