// @cap-js/db-service/lib/infer/pseudos.js

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
'use strict'

// REVISIT: we should always return cds.linked elements
// > e.g. cds.linked({definitions:{pseudos}})
const pseudos = {
  elements: {
    $user: {
      elements: {
        id: { type: 'cds.String' },
        locale: { type: 'cds.String' }, // deprecated
        tenant: { type: 'cds.String' }, // deprecated
      },
    },
    $now: { type: 'cds.Timestamp' },
    $at: { type: 'cds.Timestamp' },
    $from: { type: 'cds.Timestamp' },
    $to: { type: 'cds.Timestamp' },
    $locale: { type: 'cds.String' },
    $tenant: { type: 'cds.String' },
  },
}

module.exports = { pseudos }

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
