// tar/dist/commonjs/index.js
import __IMPORT_12__ from './write-entry.js'
import __IMPORT_11__ from './update.js'
import __IMPORT_10__ from './unpack.js'
import __IMPORT_9__ from './types.js'
import __IMPORT_8__ from './replace.js'
import __IMPORT_7__ from './read-entry.js'
import __IMPORT_6__ from './pax.js'
import __IMPORT_5__ from './parse.js'
import __IMPORT_4__ from './pack.js'
import __IMPORT_3__ from './list.js'
import __IMPORT_2__ from './header.js'
import __IMPORT_1__ from './extract.js'
import __IMPORT_0__ from './create.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.u = exports.types = exports.r = exports.t = exports.x = exports.c = void 0;
__exportStar(__IMPORT_0__(), exports);
var create_js_1 = __IMPORT_0__();
Object.defineProperty(exports, "c", { enumerable: true, get: function () { return create_js_1.create; } });
__exportStar(__IMPORT_1__(), exports);
var extract_js_1 = __IMPORT_1__();
Object.defineProperty(exports, "x", { enumerable: true, get: function () { return extract_js_1.extract; } });
__exportStar(__IMPORT_2__(), exports);
__exportStar(__IMPORT_3__(), exports);
var list_js_1 = __IMPORT_3__();
Object.defineProperty(exports, "t", { enumerable: true, get: function () { return list_js_1.list; } });
// classes
__exportStar(__IMPORT_4__(), exports);
__exportStar(__IMPORT_5__(), exports);
__exportStar(__IMPORT_6__(), exports);
__exportStar(__IMPORT_7__(), exports);
__exportStar(__IMPORT_8__(), exports);
var replace_js_1 = __IMPORT_8__();
Object.defineProperty(exports, "r", { enumerable: true, get: function () { return replace_js_1.replace; } });
exports.types = __importStar(__IMPORT_9__());
__exportStar(__IMPORT_10__(), exports);
__exportStar(__IMPORT_11__(), exports);
var update_js_1 = __IMPORT_11__();
Object.defineProperty(exports, "u", { enumerable: true, get: function () { return update_js_1.update; } });
__exportStar(__IMPORT_12__(), exports);
return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
