// @sap/cds-compiler/lib/checks/existsMustEndInAssoc.js

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
'use strict';

/**
 * A path following an “exists” predicate must always end in an association.
 *
 * @param {object} parent
 * @param {string} prop
 * @param {Array} expression
 * @param {CSN.Path} path
 */
function existsMustEndInAssoc( parent, prop, expression, path ) {
  for (let i = 0; i < expression?.length - 1; i++) {
    if (expression[i] === 'exists') {
      const next = expression[i + 1];
      const { _art } = next;
      const errorPath = path.concat([ prop, i ]);
      if (!next.SELECT && !_art?.target) {
        this.error('ref-expecting-assoc', errorPath, {
          '#': _art.type ? 'with-type' : 'std',
          elemref: next,
          type: _art.type,
        });
      }
    }
  }
}

module.exports = {
  having: existsMustEndInAssoc,
  where: existsMustEndInAssoc,
  xpr: existsMustEndInAssoc,
};

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
