// @sap/cds-compiler/lib/checks/featureFlags.js
import __IMPORT_2__ from '../transform/db/processSqlServices.js'
import __IMPORT_1__ from '../transform/featureFlags.js'
import __IMPORT_0__ from '../base/model.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
'use strict';

const { setProp } = __IMPORT_0__();
const { featureFlags } = __IMPORT_1__();
const { isSqlService, isDummyService, isDataProductService } = __IMPORT_2__();

/**
 *
 * @param {string} flag
 *
 * @returns {Function} Function to correctly set the given flag
 */
function setFeatureFlag( flag ) {
  return function setFlag() {
    if (!this.csn.meta)
      setProp(this.csn, 'meta', {});
    if (!this.csn.meta[featureFlags])
      this.csn.meta[featureFlags] = {};

    this.csn.meta[featureFlags][flag] = true;
  };
}

// Export a applyTransformations callback object that sets the feature flags if certain properties are present
module.exports = {
  value: setFeatureFlag('$calculatedElements'),
  expand: setFeatureFlag('$expandInline'),
  inline: setFeatureFlag('$expandInline'),
  kind: function setFeatureFlagForSqlService( artifact ) {
    if (isSqlService(artifact))
      setFeatureFlag( '$sqlService' ).call(this);

    if (isDummyService(artifact, this.options))
      setFeatureFlag( '$dummyService' ).call(this);

    if (isDataProductService(artifact, this.options))
      setFeatureFlag( '$dataProductService' ).call(this);
  },
};

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
