// @sap/cds-compiler/lib/checks/manyNavigations.js
import __IMPORT_0__ from '../model/csnUtils.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
'use strict';

const { applyTransformationsOnNonDictionary } = __IMPORT_0__();

/**
 * Check all refs in the given parent for the traversal of paths
 * into `.items`
 *
 * @param {object} parent Object with the expression as a property
 * @param {string} propOnParent Name of the expression property on parent
 * @param {Array} e Expression to check - see module.exports
 * @param {CSN.Path} path
 */
function navigationIntoMany( parent, propOnParent, e, path ) {
  applyTransformationsOnNonDictionary(parent, propOnParent, {
    ref: (_parent, _prop, ref, _path) => {
      const itemNavigationIndex = _parent._links?.findIndex(l => l.art.items);
      if (itemNavigationIndex !== -1 && _parent.ref.length > itemNavigationIndex + 1)
        this.message('ref-unexpected-many-navigation', _path);
    },
  }, { skipStandard: { type: true } }, path);
}

module.exports = {
  columns: navigationIntoMany,
  from: navigationIntoMany,
  on: navigationIntoMany,
  having: navigationIntoMany,
  groupBy: navigationIntoMany,
  orderBy: navigationIntoMany,
  where: navigationIntoMany,
  xpr: navigationIntoMany,
};

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
