// @sap/cds-compiler/lib/checks/utils.js
import __IMPORT_1__ from '/web/node_modules/@sap/cds-compiler/lib/transform/tupleExpansion.js'
import __IMPORT_0__ from '/web/node_modules/@sap/cds-compiler/lib/base/builtins.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
'use strict';

const { isBuiltinType } = __IMPORT_0__();
const { RelationalOperators } = __IMPORT_1__();
/**
 * Prepare the ref steps so that they are loggable
 *
 * @param {any} refStep part of a ref
 * @returns {string} Loggable string
 */
function logReady( refStep ) {
  return refStep.id || refStep;
}

/**
 * Check that the opposite operand to a relational term is something
 * structured that can be used for tuple expansion. This can either be a
 * real 'elements' thing or a managed association/composition with foreign keys.
 *
 * @TODO: This function also allows `is null` on the right-hand-side.
 *        We should move these checks to the actual tuple expansion, because
 *        if we're missing cases here, it currently results in incorrect expansion.
 *
 * @param {Array} on the on condition which to check
 * @param {number} startIndex the index of the relational term in the on condition array
 * @returns {boolean} indicates whether the other side of a relational term is expandable
 */
function otherSideIsExpandableStructure( on, startIndex ) {
  if (on[startIndex - 1] && RelationalOperators.includes(on[startIndex - 1])) {
    const lhs = on[startIndex - 2];
    return isOk(resolveArtifactType.call(this, lhs?._art));
  }
  else if (on[startIndex + 1] && RelationalOperators.includes(on[startIndex + 1])) {
    const op = on[startIndex + 1];
    const rhs = on[startIndex + 2];
    if (op === 'is')
      // check for unary operator 'is [not] null' as token stream
      return rhs === 'null' || (rhs === 'not' && on[startIndex + 3] === 'null');
    return isOk(resolveArtifactType.call(this, rhs?._art));
  }
  return false;

  /**
   * Artifact is structured or a managed association/composition
   *
   * @param {CSN.Artifact} art Artifact
   * @returns {boolean} True if expandable
   */
  function isOk( art ) {
    return !!(art && (art.elements || (art.target && art.keys)));
  }
}

/**
 * Check that the opposite operand to a relational term is s value or "is null".
 *
 * @param {Array} expr the expression which to check
 * @param {number} startIndex the index of the relational term in the expression array
 * @returns {boolean} indicates whether the other side of a relational term is scalar
 */
function otherSideIsValue( expr, startIndex ) {
  if (expr[startIndex - 1] && RelationalOperators.includes(expr[startIndex - 1]))
    return expr[startIndex - 2]?.val !== undefined;

  if (expr[startIndex + 1] && RelationalOperators.includes(expr[startIndex + 1])) {
    const op = expr[startIndex + 1];
    const rhs = expr[startIndex + 2];
    if (op === 'is')
      // check for unary operator 'is [not] null' as token stream
      return rhs === 'null' || (rhs === 'not' && expr[startIndex + 3] === 'null');
    return rhs?.val !== undefined;
  }

  return false;
}

/**
 * Get the real type of an artifact
 *
 * @param {object} art Whatever _art by csnRefs can be - element or artifact
 * @returns {object} final artifact type
 */
function resolveArtifactType( art ) {
  const type = art?._type?.type || art?.type;
  if (type && !isBuiltinType(type))
    return this.csnUtils.getFinalTypeInfo(type);

  return art;
}

module.exports = {
  logReady,
  otherSideIsExpandableStructure,
  otherSideIsValue,
  resolveArtifactType,
};

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
