// @sap/cds-compiler/lib/checks/validator.js
import __IMPORT_30__ from '../utils/timetrace.js'
import __IMPORT_29__ from './featureFlags.js'
import __IMPORT_28__ from './assocOutsideService.js'
import __IMPORT_27__ from './sql-snippets.js'
import __IMPORT_26__ from './managedWithoutKeys.js'
import __IMPORT_25__ from './checkPathsInStoredCalcElement.js'
import __IMPORT_24__ from './existsExpressionsOnlyForeignKeys.js'
import __IMPORT_23__ from './existsMustEndInAssoc.js'
import __IMPORT_22__ from './nullableKeys.js'
import __IMPORT_21__ from './queryNoDbArtifacts.js'
import __IMPORT_20__ from './arrayOfs.js'
import __IMPORT_19__ from './invalidTarget.js'
import __IMPORT_18__ from './elements.js'
import __IMPORT_17__ from './types.js'
import __IMPORT_16__ from './foreignKeys.js'
import __IMPORT_15__ from './onConditions.js'
import __IMPORT_14__ from './cdsMap.js'
import __IMPORT_13__ from './annotationsOData.js'
import __IMPORT_12__ from './actionsFunctions.js'
import __IMPORT_11__ from './enums.js'
import __IMPORT_10__ from './parameters.js'
import __IMPORT_9__ from './structuredAnnoExpressions.js'
import __IMPORT_8__ from './checkForTypes.js'
import __IMPORT_7__ from './hasPersistedElements.js'
import __IMPORT_6__ from './managedInType.js'
import __IMPORT_5__ from './manyNavigations.js'
import __IMPORT_4__ from './cdsPersistence.js'
import __IMPORT_3__ from './defaultValues.js'
import __IMPORT_2__ from './selectItems.js'
import __IMPORT_1__ from './enricher.js'
import __IMPORT_0__ from '../model/csnUtils.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
'use strict';

const {
  forEachDefinition, forEachMemberRecursively, forAllQueries,
  forEachMember, getNormalizedQuery,
  applyTransformations, functionList, mergeTransformers, hasPersistenceSkipAnnotation,
} = __IMPORT_0__();
const enrichCsn = __IMPORT_1__();

// forRelationalDB
const { validateSelectItems } = __IMPORT_2__();
const { rejectParamDefaultsInHanaCds, warnAboutDefaultOnAssociationForHanaCds } = __IMPORT_3__();
const validateCdsPersistenceAnnotation = __IMPORT_4__();
const navigationIntoMany = __IMPORT_5__();
const checkUsedTypesForAnonymousAspectComposition = __IMPORT_6__();
const validateHasPersistedElements = __IMPORT_7__();
const checkForHanaTypes = __IMPORT_8__();
const { checkAnnotationExpression } = __IMPORT_9__();
const checkForParams = __IMPORT_10__();
const checkAndRemoveEnums = __IMPORT_11__();
// forOdata
const { validateDefaultValues } = __IMPORT_3__();
const { checkActionOrFunction } = __IMPORT_12__();
const {
  checkCoreMediaTypeAllowance, checkAnalytics,
  checkAtSapAnnotations, checkReadOnlyAndInsertOnly,
  checkTemporalAnnotationsAssignment,
} = __IMPORT_13__();
// both
const checkCdsMap = __IMPORT_14__();
const { validateOnCondition, validateMixinOnCondition } = __IMPORT_15__();
const validateForeignKeys = __IMPORT_16__();
const {
  checkTypeDefinitionHasType, checkElementTypeDefinitionHasType,
  checkTypeIsScalar, checkDecimalScale,
} = __IMPORT_17__();
const {
  checkPrimaryKey, checkVirtualElement, checkManagedAssoc,
  checkRecursiveTypeUsage, rejectAnnotationsOnCalcElement,
} = __IMPORT_18__();
const checkForInvalidTarget = __IMPORT_19__();
const { validateAssociationsInItems } = __IMPORT_20__();
const checkQueryForNoDBArtifacts = __IMPORT_21__();
const checkExplicitlyNullableKeys = __IMPORT_22__();
const existsMustEndInAssoc = __IMPORT_23__();
const forbidAssocInExists = __IMPORT_24__();
const checkPathsInStoredCalcElement = __IMPORT_25__();
const managedWithoutKeys = __IMPORT_26__();
const {
  checkSqlAnnotationOnArtifact,
  checkSqlAnnotationOnElement,
} = __IMPORT_27__();
const assertNoAssocUsageOutsideOfService = __IMPORT_28__();
const featureFlags = __IMPORT_29__();
const { timetrace } = __IMPORT_30__();

const forRelationalDBMemberValidators
  = [
  // For HANA CDS specifically, reject any default parameter values, as these are not supported.
    rejectParamDefaultsInHanaCds,
    checkTypeIsScalar,
    checkDecimalScale,
    checkExplicitlyNullableKeys,
    managedWithoutKeys,
    warnAboutDefaultOnAssociationForHanaCds,
    // sql.prepend/append
    checkSqlAnnotationOnElement,
    // no temporal annotations on calc elements
    rejectAnnotationsOnCalcElement,
    checkElementTypeDefinitionHasType,
  ];

const forRelationalDBArtifactValidators = [
  checkPrimaryKey,
  // @cds.persistence has no impact on odata
  validateCdsPersistenceAnnotation,
  // virtual items are not persisted on the db
  validateHasPersistedElements,
  // sql.prepend/append
  checkSqlAnnotationOnArtifact,
  // strip down CSN to reduce it's size by removing non-sql relevant parts
  deleteBoundActions,
];

const forRelationalDBCsnValidators = [
  checkCdsMap,
  existsMustEndInAssoc,
  forbidAssocInExists,
  navigationIntoMany,
  checkPathsInStoredCalcElement,
  featureFlags,
  checkAndRemoveEnums,
];
/**
 * @type {Array<(query: CSN.Query, path: CSN.Path) => void>}
 */
const forRelationalDBQueryValidators = [
  // TODO reason why this is forRelationalDB exclusive
  validateSelectItems,
  checkQueryForNoDBArtifacts,
];

const forOdataMemberValidators
  = [
  // OData allows only simple values, no expressions or functions
    validateDefaultValues,
    managedWithoutKeys,
  ];

const forOdataArtifactValidators
  = [
  // actions and functions are not of interest for the database
    checkActionOrFunction,
    // arrays are just CLOBs/LargeString for the database,
    // no inner for the array structure is of interest for the database
    // NOTE: moved to the renderer for a while
    // TODO: Re-enable this code and remove the duplicated code from the renderer.
    //       Not possible at the moment, because running this at the beginning of
    //       the renderer does not work because the enricher can't handle certain
    //       OData specifics.
    // checkChainedArray,
    checkReadOnlyAndInsertOnly,
  ];

const forOdataCsnValidators = [ checkCdsMap ];

const forOdataQueryValidators = [];

const commonMemberValidators
  = [ validateOnCondition, validateForeignKeys,
    validateAssociationsInItems, checkForInvalidTarget,
    checkVirtualElement, checkManagedAssoc ];

// TODO: checkManagedAssoc is a forEachMemberRecursively!
const commonArtifactValidators = [
  checkTypeDefinitionHasType,
  checkRecursiveTypeUsage,
];
// TODO: Does it make sense to run the on-condition check as part of a CSN validator?
const commonQueryValidators = [ validateMixinOnCondition ];

/**
 * Run the given validations for each artifact and for each member recursively
 *
 * @param {CSN.Model} csn CSN to check
 * @param {object} that Will be provided to the validators via "this"
 * @param {object[]} [csnValidators] Validations on whole CSN using applyTransformations
 * @param {Function[]} [memberValidators] Validations on member-level
 * @param {Function[]} [artifactValidators] Validations on artifact-level
 * @param {Function[]} [queryValidators] Validations on query-level
 * @param {object} iterateOptions can be used to skip certain kinds from being iterated e.g. 'action' and 'function' for hana
 * @returns {Function} Function taking no parameters, that cleans up the attached helpers
 */
function _validate( csn, that,
                    csnValidators = [],
                    memberValidators = [],
                    artifactValidators = [],
                    queryValidators = [],
                    iterateOptions = {} ) {
  timetrace.start('Enrich CSN');
  const { cleanup } = enrichCsn(csn, that.options);
  timetrace.stop('Enrich CSN');
  // TODO: Don't know if that's feasible? Do we really need to enrich annotations always?
  // const { cleanup } = enrich(csn, { processAnnotations: that.options.tranformation === 'odata' });

  applyTransformations(csn, mergeTransformers(csnValidators, that), [], { drillRef: true });

  forEachDefinition(csn, (artifact, artifactName, prop, path) => {
    artifactValidators.forEach((artifactValidator) => {
      artifactValidator.bind(that)(artifact, artifactName, prop, path);
    });
    that.artifact = artifact;
    if (memberValidators.length) {
      forEachMemberRecursively( artifact,
                                memberValidators.map(v => v.bind(that)),
                                path,
                                true,
                                iterateOptions );
    }

    if (queryValidators.length && getNormalizedQuery(artifact).query) {
      forAllQueries(getNormalizedQuery(artifact).query, functionList(queryValidators, that),
                    path.concat([ artifact.projection ? 'projection' : 'query' ]));
    }
  }, iterateOptions);

  return cleanup;
}

/**
 * Depending on the dialect we need to run different validations.
 *
 * @param {CSN.Options} options
 * @returns {any[]} Array of validator functions (or objects?)
 */
function getDBCsnValidators( options ) {
  const validations = [ ...forRelationalDBCsnValidators ];

  if (options.transformation !== 'effective')
    validations.push(checkForParams.csnValidator);
  if (options.sqlDialect === 'h2' || options.sqlDialect === 'postgres')
    validations.push(checkForHanaTypes);

  return validations;
}

/**
 * @param {CSN.Model} csn CSN to check
 * @param {object} that Will be provided to the validators via "this"
 * @returns {Function} the validator function with the respective checks for the HANA backend
 */
function forRelationalDB( csn, that ) {
  const memberValidators = [ ...forRelationalDBMemberValidators, ...commonMemberValidators ];
  if (that.options.transformation === 'hdbcds')
    memberValidators.push(checkForParams.memberValidator);

  return _validate(csn, that,
                   getDBCsnValidators(that.options),
                   memberValidators,
                   forRelationalDBArtifactValidators.concat(commonArtifactValidators).concat(
                     // why is this hana exclusive
                     (artifact) => {
                       /*  the validation itself performs a recursive check on structured elements.
                        That is why it is not run along with the memberValidators, as it would result in
                        duplicate messages due to the forEachMemberRecursively.
                        TODO: check if this recursion can be factored out of the validator */
                       forEachMember(artifact, checkUsedTypesForAnonymousAspectComposition.bind(that));
                     },
                     (artifact, artifactName) => {
                       if (that.options.transformation === 'effective') {
                         if (that.options.effectiveServiceName)
                           assertNoAssocUsageOutsideOfService.bind(that)(artifact, artifactName);

                         forEachMemberRecursively(artifact, checkAnnotationExpression.bind(that), [ 'definitions', artifactName ], false, {
                           skipArtifact: a => a.returns || (a.params && !a.query),
                         });
                       }
                     }
                   ),
                   forRelationalDBQueryValidators.concat(commonQueryValidators),
                   {
                     skipArtifact: artifact => artifact.abstract ||
                                                hasPersistenceSkipAnnotation(artifact) ||
                                                artifact['@cds.persistence.exists'] ||
                                                [ 'action', 'function', 'event' ].includes(artifact.kind),
                   });
}

/**
 * @param {CSN.Model} csn CSN to check
 * @param {object} that Will be provided to the validators via "this"
 * @returns {Function} the validator function with the respective checks for the OData backend
 */
function forOdata( csn, that ) {
  return _validate(csn, that,
                   forOdataCsnValidators,
                   forOdataMemberValidators.concat(commonMemberValidators),
                   forOdataArtifactValidators.concat(commonArtifactValidators).concat(
                     (artifact, artifactName) => {
                       if (that.csnUtils.getServiceName(artifactName)) {
                         checkAtSapAnnotations.bind(that)(artifact);
                         forEachMemberRecursively(artifact, [
                           checkCoreMediaTypeAllowance.bind(that),
                           checkAnalytics.bind(that),
                           checkAtSapAnnotations.bind(that),
                         ]);
                       }
                       checkTemporalAnnotationsAssignment.bind(that)(artifact, artifactName);
                     }
                   ),
                   // eslint-disable-next-line sonarjs/no-empty-collection
                   forOdataQueryValidators.concat(commonQueryValidators),
                   {
                     skipArtifact: this.isExternalServiceMember,
                   });
}

/**
 * Shrink the CSN by
 * - deleting bound actions
 *
 * Artifacts can only be shrunk later, when types are resolved
 *
 * @param {CSN.Artifact} artifact
 */
function deleteBoundActions( artifact ) {
  if (this.options.transformation !== 'effective')
    delete artifact.actions;
}

module.exports = { forRelationalDB, forOdata };

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
