// @sap/cds/libx/_runtime/remote/Service.js
import __IMPORT_0__ from '../../../lib/index.js'

const require = function () { throw Object.assign(new Error('require in browser'), { code: 'MODULE_NOT_FOUND' }) }
require.resolve = function (id) { return new URL(id, import.meta.url).pathname }
export default (function (exports, module, require, __dirname, __filename) {

  const cds = __IMPORT_0__()

  class RemoteService extends cds.Service {
    init() {
      if (([...this.entities].length || [...this.operations].length) && !this.options.credentials)
        cds.error`No credentials configured for "${this.name}".`

      this.kind = this.options.kind
      this.url = this.options.credentials.url
      if (!this.url.endsWith('/')) this.url = `${this.url}/`

      // Add 'initial' handler to clear keys from data
      const clearKeysFromData = function (req) {
        if (req.target && req.target.keys) for (const k of Object.keys(req.target.keys)) delete req.data[k]
      }
      this.before('UPDATE', '*', Object.assign(clearKeysFromData, { _initial: true }))

      this.on('*', async function on_handler(req, next) {
        const { method, headers, query } = req
        if (!query && !(typeof req.path === 'string')) return next()

        if (this.online === false || navigator.onLine === false) throw new Error(`Failed to connect to: ${this.url}`)

        try {
          const url = cds.odata.urlify(query, { kind: this.kind, model: this.model, method })
          const response = await fetch(this.url + url.path, { method: url.method, headers: { 'x-correlation-id': cds.context.id, ...headers } })
          let result = await response.json()
          result = result.value ?? result
          result = typeof query === 'object' && query.SELECT?.one && Array.isArray(result) ? result[0] : result
          return result
        } catch (err) {
          if (err instanceof TypeError) this.__proto__.online = false
          throw err
        }
      })

      return super.init()
    }
  }

  RemoteService.prototype.isExternal = true
  module.exports = RemoteService

  return module.exports
}).bind(globalThis, {}, {}, require, new URL('.', import.meta.url).pathname, new URL('', import.meta.url).pathname)
