// @sap/cds/libx/_runtime/common/Service.js
import __IMPORT_13__ from 'generic/auth/restrictions.js'
import __IMPORT_12__ from 'generic/assert.js'
import __IMPORT_11__ from 'generic/flows.js'
import __IMPORT_10__ from 'generic/crud.js'
import __IMPORT_9__ from '../fiori/lean-draft.js'
import __IMPORT_8__ from 'generic/sorting.js'
import __IMPORT_7__ from 'generic/paging.js'
import __IMPORT_6__ from 'generic/temporal.js'
import __IMPORT_5__ from 'generic/stream-only.js'
import __IMPORT_4__ from 'generic/input.js'
import __IMPORT_3__ from 'generic/etag.js'
import __IMPORT_2__ from 'generic/auth/index.js'
import __IMPORT_1__ from 'utils/postProcess.js'
import __IMPORT_0__ from '../cds.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
const cds = __IMPORT_0__()

const postProcess = __IMPORT_1__()

/**
 * Generic Application Service Provider
 */
class ApplicationService extends cds.Service {
  init() {
  const clazz = this.constructor
    for (let each of clazz.generics) clazz[each].call(this)
    return super.init()
  }

  static get generics() {
    return (this._generics ??= new Set([
      ...(this.__proto__.generics || []),
      ...Reflect.ownKeys(this).filter(p => p.startsWith('handle_'))
    ]))
  }

  static get handle_authorization() {
    return __IMPORT_2__()
  }

  static get handle_etags() {
    return __IMPORT_3__()
  }

  static get handle_validations() {
    return __IMPORT_4__()
  }

  static get handle_stream_property() {
    return __IMPORT_5__()
  }

  static get handle_temporal_data() {
    return __IMPORT_6__()
  }

  static get handle_paging() {
    return __IMPORT_7__() // > paging must be executed before sorting
  }

  static get handle_sorting() {
    return __IMPORT_8__()
  }

  static get handle_drafts() {
    return __IMPORT_9__()
  }

  static get handle_crud() {
    return __IMPORT_10__()
  }

  static get handle_flows() {
    return __IMPORT_11__()
  }

  static get handle_assert() {
    return __IMPORT_12__()
  }

  // Overload .handle in order to resolve projections up to a definition that is known by the remote service instance.
  // Result is post processed according to the inverse projection in order to reflect the correct result of the original query.
  async handle(req) {
    // REVISIT: We must not allow arbitrary CQNs, so this shouldn't be here!
    if (!this._requires_resolving?.(req)) return super.handle(req)
    // rewrite the query to a target entity served by this service...
    const query = this.resolve(req.query)
    if (!query) cds.error`Target ${req.target.name} cannot be resolved for service ${this.name}`
    const target = query._target || req.target
    // we need to provide target explicitly because it's cached within ensure_target
    const _req = new cds.Request({ query, target, _resolved: true })
    return await this.dispatch(_req).then(result => postProcess(query, result, this))
  }
}

// NOTE: getRestrictions is VERY INOFFICIAL!!!
//> only kept temporary because of an ill-devised usage in AMS plugin
const { getRestrictions } = __IMPORT_13__()
ApplicationService.prototype.getRestrictions = function (..._) {
  return getRestrictions.call(this, ..._)
}

ApplicationService.prototype.isAppService = true
module.exports = ApplicationService

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
