// @sap/cds/lib/utils/tar-lib.js
import __IMPORT_3__ from '/web/node/stream/consumers.mjs'
import __IMPORT_2__ from '/web/node/stream.mjs'
import __IMPORT_1__ from '/web/node_modules/tar/dist/commonjs/index.js'
import __IMPORT_0__ from '/web/node_modules/@sap/cds/lib/index.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
const cds = __IMPORT_0__(), { path, mkdirp } = cds.utils
const tar = __IMPORT_1__()
const { Readable } = __IMPORT_2__
const cons = __IMPORT_3__

const _resolve = (...x) => path.resolve (cds.root,...x)

exports.create = async (root, ...args) => {  
  if (typeof root === 'string') root = _resolve(root)
  if (Array.isArray(root)) [ root, ...args ] = [ cds.root, root, ...args ]

  const options = {}
  if (args.includes('-z')) options.gzip = true
  const index = args.findIndex(el => el === '-f')
  if (index>=0) options.file = _resolve(args[index+1])
  options.cwd = root
    
  let dirs = []
  for (let i=0; i<args.length; i++) {
    if (args[i] === '-z' || args[i] === '-f') break
    if (Array.isArray(args[i])) args[i].forEach(a => dirs.push(_resolve(a)))
    else if (typeof args[i] === 'string') dirs.push(_resolve(args[i]))
  }
  if (!dirs.length) dirs.push(root)
  dirs = dirs.map(d => path.relative(root, d))
  
  const stream = await tar.c(options, dirs)

  return stream && await cons.buffer(stream) 
}

exports.extract = (archive, ...args) => ({
  async to (dest) {
    if (typeof dest === 'string') dest = _resolve(dest)
    const stream = Readable.from(archive)
    
    const options = { C: dest }
    if (args.includes('-z')) options.gzip = true

    return new Promise((resolve, reject) => {
        const tr = tar.x(options)
        stream.pipe(tr)
        tr.on('close', () => resolve())
        tr.on('error', e => reject(e))
    })
  }
})

const tar_ = exports
exports.c = tar_.create
exports.cz = (d,...args) => tar_.c (d, ...args, '-z')
exports.cf = (t,d,...args) => tar_.c (d, ...args, '-f',t)
exports.czf = (t,d,...args) => tar_.c (d, ...args, '-z', '-f',t)
exports.czfd = (t,...args) => mkdirp(path.dirname(t)).then (()=> tar_.czf (t,...args))
exports.x = tar_.xf = tar_.extract
exports.xz = tar_.xzf = a => tar_.x (a, '-z')
exports.xv = tar_.xvf = a => tar_.x (a)
exports.xvz = tar_.xvzf = a => tar_.x (a, '-z')

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
