// js-yaml/index.js
import __IMPORT_21__ from './lib/type/str.js'
import __IMPORT_20__ from './lib/type/seq.js'
import __IMPORT_19__ from './lib/type/omap.js'
import __IMPORT_18__ from './lib/type/merge.js'
import __IMPORT_17__ from './lib/type/int.js'
import __IMPORT_16__ from './lib/type/bool.js'
import __IMPORT_15__ from './lib/type/timestamp.js'
import __IMPORT_14__ from './lib/type/set.js'
import __IMPORT_13__ from './lib/type/pairs.js'
import __IMPORT_12__ from './lib/type/null.js'
import __IMPORT_11__ from './lib/type/map.js'
import __IMPORT_10__ from './lib/type/float.js'
import __IMPORT_9__ from './lib/type/binary.js'
import __IMPORT_8__ from './lib/exception.js'
import __IMPORT_7__ from './lib/schema/default.js'
import __IMPORT_6__ from './lib/schema/core.js'
import __IMPORT_5__ from './lib/schema/json.js'
import __IMPORT_4__ from './lib/schema/failsafe.js'
import __IMPORT_3__ from './lib/schema.js'
import __IMPORT_2__ from './lib/type.js'
import __IMPORT_1__ from './lib/dumper.js'
import __IMPORT_0__ from './lib/loader.js'

const require = function() {throw Object.assign(new Error('require in browser'),{code:'MODULE_NOT_FOUND'})}
require.resolve = function(id) { 
  if(id[0] === '.') return new URL(id, import.meta.url).pathname
  throw Object.assign(new Error('require.resolve in browser'),{ code: 'MODULE_NOT_FOUND' }) 
}
export default (function(exports, module, require, __dirname,__filename) {
if(module.__loaded__) { return module.exports }
module.__loaded__ = true
module.exports = exports;
'use strict';


var loader = __IMPORT_0__();
var dumper = __IMPORT_1__();


function renamed(from, to) {
  return function () {
    throw new Error('Function yaml.' + from + ' is removed in js-yaml 4. ' +
      'Use yaml.' + to + ' instead, which is now safe by default.');
  };
}


module.exports.Type                = __IMPORT_2__();
module.exports.Schema              = __IMPORT_3__();
module.exports.FAILSAFE_SCHEMA     = __IMPORT_4__();
module.exports.JSON_SCHEMA         = __IMPORT_5__();
module.exports.CORE_SCHEMA         = __IMPORT_6__();
module.exports.DEFAULT_SCHEMA      = __IMPORT_7__();
module.exports.load                = loader.load;
module.exports.loadAll             = loader.loadAll;
module.exports.dump                = dumper.dump;
module.exports.YAMLException       = __IMPORT_8__();

// Re-export all types in case user wants to create custom schema
module.exports.types = {
  binary:    __IMPORT_9__(),
  float:     __IMPORT_10__(),
  map:       __IMPORT_11__(),
  null:      __IMPORT_12__(),
  pairs:     __IMPORT_13__(),
  set:       __IMPORT_14__(),
  timestamp: __IMPORT_15__(),
  bool:      __IMPORT_16__(),
  int:       __IMPORT_17__(),
  merge:     __IMPORT_18__(),
  omap:      __IMPORT_19__(),
  seq:       __IMPORT_20__(),
  str:       __IMPORT_21__()
};

// Removed functions from JS-YAML 3.0.x
module.exports.safeLoad            = renamed('safeLoad', 'load');
module.exports.safeLoadAll         = renamed('safeLoadAll', 'loadAll');
module.exports.safeDump            = renamed('safeDump', 'dump');

return module.exports
}).bind(globalThis,{},{},require,new URL('.', import.meta.url).pathname,new URL('', import.meta.url).pathname)
